src/rgw/rgw_acl.h

```diff
@@ -215,24 +215,56 @@ struct ACLReferer {
   }
 
   bool is_match(std::string http_referer) const {
-    if (http_referer == url_spec) {
+    std::string http_host;
+    if (!get_http_host(http_referer, http_host))
+      return false;
+
+    if (http_host == url_spec) {
       return true;
     }
 
-    if (http_referer.length() < url_spec.length()) {
+    if (http_host.length() < url_spec.length()) {
       return false;
     }
 
     if ('.' == url_spec[0]) {
       /* Wildcard support: a referer matches the spec when its last char are
        * perfectly equal to spec. */
-      return !http_referer.compare(http_referer.length() - url_spec.length(),
+      return !http_host.compare(http_host.length() - url_spec.length(),
                                    url_spec.length(), url_spec);
     }
 
     return false;
   }
 
+  bool get_http_host(std::string url, std::string& http_host) const {
+    int pos = url.find("://");
+    if (pos < 0 || pos == (int)url.size() - 3)
+      return false;
+
+    string scheme = url.substr(0, pos);
+    if (scheme.empty())
+      return false;
+    
+    string url_sub = url.substr(pos + 3);
+    pos = url_sub.find('@');
+    if (pos == (int)url_sub.size())
+      return false;
+
+    url_sub = url_sub.substr(pos + 1);
```

Comment:
If `url_sub` doesn't contain `@` then we're doing `std::string::npos + 1` which results to `0` on most implementations. However, this isn't a way to go.